### ISSUE TITLE
--sort does not sort entries with = assignments correctly

### ISSUE THREAD

**Issue by JosefAtmin** [ledger/ledger#2065]
Hello,

the --sort option does not sort the entries as expected.  Seems to be a bug to me.

I have the file:
=========================================================================================

2021-01-01 A
    account1                                                              100.00 EUR
    account2

2021-04-01 D
    account1                                                                          = 300.00 EUR
    account2

2021-02-01 B
    account1                                                               10.00 EUR
    account2

2021-03-01 C
    account1                                                                          = 200.00 EUR
    account2

=========================================================================================

I run the command:
ledger --date-format '%Y-%m-%d' -f File.ledger --sort-all 'date' reg account1

I get the output:
=========================================================================================
2021-01-01 A                                  account1                                          100.00 EUR           100.00 EUR
2021-02-01 B                                  account1                                           10.00 EUR           110.00 EUR
2021-03-01 C                                  account1                                         -110.00 EUR                    0
2021-04-01 D                                  account1                                          200.00 EUR           200.00 EUR
=========================================================================================

The expected output is:
=========================================================================================
2021-01-01 A                                  account1                                          100.00 EUR           100.00 EUR
2021-02-01 B                                  account1                                           10.00 EUR           110.00 EUR
2021-03-01 C                                  account1                                           90.00 EUR           200.00 EUR
2021-04-01 D                                  account1                                          100.00 EUR           300.00 EUR
=========================================================================================

hledger produces the expected output, but I would prefer to use ledger.

I find it very strange that ledger expects the entries to be sorted in the file in the first place.  Feels like a severe constraint.  So I was glad to see the --sort option.  What a pity that it does not work (as expected).

Thanks for the great program anyways.

Best, Josef.

**Comment by tbm** [comment]
If I understand this bug report correctly, I find it very misleading.  Yes, you're not getting the result you want, but the actual sorting is done correctly.  You want to sort by date and it sorts by date.

As you point out, ledger processes balance assignments in the order of transactions in the file, not based on the date.  This is arguably a bad idea and should be fixed (and there are bug reports about this behaviour), but given this behaviour the result you get is consistent:

Looking at the transactions in *file order*, we have:

* 100 EUR
* Setting to 300 EUR, i.e. an adjustment of +200
* +10 (i.e. 310 now)
* Setting to 200, i.e. an adjustment of -110

And you use `--sort` to sort by date, but that doesn't change that the balance assignments are processed based on file order.

So the result you get is consistent.

Yes, it's not what you want and I'd personally agree it's not what makes sense.  But it is completely consistent with what ledger is doing and the actual sorting is correct (you get transactions according to dates).

In other words, `--sort` only influences how transactions are shown, but not how they are processed.  Maybe that wasn't clear to you.

**Comment by JosefAtmin** [comment]
Thank you very much for this quick and clear clarification. I understand. But that really rules out ledger for me, and I will probably use hledger now. I thought that --sort fixes this bad design decision. It would be great if there were a sort option that would apply before processing the transactions.

**Comment by jwiegley** [comment]
The design decision in Ledger is that data has a single meaning, while what is reported can very. If the "meaning" of balance assignments were based on a reporting option, like sorting, it would invalidate this core principle. I would not call this a bad design decision, but rather a distinction between the meaning of data (which has to decide on a fixed order in order for syntactic constructs to have a universal meaning), and the meaning of what is displayed — which should always depend entirely on the data, and not some combination of reporting flags, which in Ledger can come from many sources (command-line, `~/.ledgerrc`, included files, etc).
